### ISSUE TITLE
Controller doesn't reconcile to changes to application workloads

### ISSUE THREAD

**Issue by cdlliuy** [redhat-developer/service-binding-operator#422]
```
spec:
  applicationSelector:
    group: serving.knative.dev
    resource: services
    resourceRef: mydemo
    version: v1
.....
```
Once the sbr created,  I tried to delete `mydemo` , then create it again with the same name, but  SBR didn't  reconcile to have the secrets injected into the new `mydemo` container anymore.   

Similar to backingEndServiceSelectors as well,  I delete a specific backingservice, then create it again with the same name but different secret content.  Anyway, the new secret didn't inject into the application container. 

In short:
1.  create sbr successfully
2.  delete application  
3.  create application again with the same name.  Keep SBR unchanged
4.   no env injected in the new application container. 

OR
1.  create sbr successfully
2.  delete the backingservice  
3.  create backingservice again with the same name.  Keep SBR unchanged
4.  With the new backingservice, the service credential is changed.   But the injected env in application containers is not updated.

**Comment by sbose78** [comment]
Thanks for reporting this.

@isutton @baijum @DhritiShikhar @Avni-Sharma  could you please take a look?

**Comment by DhritiShikhar** [comment]
Yes @sbose78 I am looking into this..

**Comment by cdlliuy** [comment]
@DhritiShikhar  Thanks and eager to know your findings :-)

**Comment by int8dev** [comment]
AFAIK currently operator only watch the following resources

- ServiceBindingRequest CRD 
- Configmaps
- Secrets
- CRDs owned by ClusterServiceVersion resources.

So If we specify other CRDs in `applicationSelector` or `backingServiceSelectors`, reconcile would not be triggered as they are not watched by informer. 

And I also noticed another potential issue that after a success reconcile for knative service, there are no SBR informations in its annotations field. So I think even we added operator to watch knative services the reconcile will not working as expected. 

Checking knative service after creating SBR successsfully. No annotations for SBR information.
```
apiVersion: serving.knative.dev/v1
kind: Service
metadata:
  annotations:
    serving.knative.dev/lastModifier: system:serviceaccount:ibm-service-binding:service-binding-operator
  creationTimestamp: "2020-04-22T13:31:11Z"
  generation: 3
  name: helloworld
  namespace: 930852e2-7e33
  resourceVersion: "48100370"
  selfLink: /apis/serving.knative.dev/v1/namespaces/930852e2-7e33/services/helloworld
  uid: 177a47b2-04e7-45de-837d-1081ccd37e5d
spec:
  template:
    metadata:
      annotations:
        client.knative.dev/user-image: gcr.io/knative-samples/helloworld-go
      creationTimestamp: null
    spec:
      containerConcurrency: 0
      containers:
      - env:
        - name: ServiceBindingOperatorChangeTriggerEnvVar
          value: "2020-04-22T13:45:44Z"
        envFrom:
        - secretRef:
            name: helloworld-bindingrequest
        image: gcr.io/knative-samples/helloworld-go
        name: user-container
        readinessProbe:
          successThreshold: 1
          tcpSocket:
            port: 0
        resources: {}
      timeoutSeconds: 300
  traffic:
  - latestRevision: true
    percent: 100
```

Checking backend service binding to the ksvc. Annotations for SBR are present.
```
apiVersion: ibmcloud.ibm.com/v1alpha1
kind: Binding
metadata:
  annotations:
    service-binding-operator.apps.openshift.io/binding-name: helloworld-bindingrequest
    service-binding-operator.apps.openshift.io/binding-namespace: 930852e2-7e33
  creationTimestamp: "2020-04-09T13:26:08Z"
  finalizers:
  - binding.ibmcloud.ibm.com
  generation: 1
  name: my-language-translator-credential
  namespace: 930852e2-7e33
  ownerReferences:
  - apiVersion: ibmcloud.ibm.com/v1alpha1
    blockOwnerDeletion: true
    controller: true
    kind: Service
    name: my-language-translator
    uid: e89ccbf3-3ecb-4514-a54e-373c50c4d5
```

**Comment by sbose78** [comment]
The controller doesn't "manage"/"watch" the workload at the moment and it was a design decision to let the admins be the ultimate manager of workloads. I am happy to revisit it - hence there are no annotations on deployments, DeploymentConfigs, KnativeServices.

However, on modification of the SBR, a reconcile should be triggered, that sounds like a bug.

**Comment by sbose78** [comment]
@cdlliuy Did the initial binding work end-to-end ?

1. Create an app
2. Create a backing service
3. Create an SBR

**Comment by cdlliuy** [comment]
@sbose78 ,  yes , followed the above sequence to create sbr.

**Comment by int8dev** [comment]
I think we have several options to address the reconcile issue.
1. Add informer to watch whatever CRD defined in the applicationSelector. This might cause overhead as some resources like knative service are designed to update frequently and adding too many resource kinds to watch would cause potential performance issues. It would be even worse if the service binding operator is running at cluster level.
2. Add a field in applicationSelector or somewhere else in SBR CRD to determin whether the operator should watch this kind of resource. This gives flexibility to watch certain kinds of  resource.
3. Make the watching GVKs configurable when starting service binding operator. This is similar to 2) but with less flexibility.

**Comment by qibobo** [comment]
Below is how I did the test:
**1 create a kn service:**

```
apiVersion: serving.knative.dev/v1
kind: Service 
metadata:
  name: mytranslator-web-knative-vcap6
spec:
  template:
    metadata:
      annotations:
        autoscaling.knative.dev/minScale: "1"
    spec:
      containers:
        - image: docker.io/qibobo/go4translator_vcap:latest
          ports:
          - containerPort: 8080
```
**2 create a SBR:**

```
apiVersion: apps.openshift.io/v1alpha1
kind: ServiceBindingRequest
metadata:
  name: binding-request-vcap6
spec:
  envVarPrefix: prefix1
  applicationSelector:
    resourceRef: mytranslator-web-knative-vcap6
    group: serving.knative.dev 
    version: v1 
    resource: services 
  backingServiceSelectors:
    - group: ibmcloud.ibm.com
      version: v1alpha1
      kind: Binding
      resourceRef: thetranslator-service-key
```
And then I can find the secret is injected:

```
kubectl get ksvc mytranslator-web-knative-vcap6 -o yaml
apiVersion: serving.knative.dev/v1
kind: Service
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"serving.knative.dev/v1","kind":"Service","metadata":{"annotations":{},"name":"mytranslator-web-knative-vcap6","namespace":"e837ea57-f035"},"spec":{"template":{"metadata":{"annotations":{"autoscaling.knative.dev/minScale":"1"}},"spec":{"containers":[{"image":"docker.io/qibobo/go4translator_vcap:latest","ports":[{"containerPort":8080}]}]}}}}
    serving.knative.dev/creator: system:serviceaccount:e837ea57-f035:e837ea57-f035-writer
    serving.knative.dev/lastModifier: system:serviceaccount:ibm-service-binding:service-binding-operator
  creationTimestamp: "2020-04-23T09:15:28Z"
  generation: 2
  name: mytranslator-web-knative-vcap6
  namespace: e837ea57-f035
  resourceVersion: "48636092"
  selfLink: /apis/serving.knative.dev/v1/namespaces/e837ea57-f035/services/mytranslator-web-knative-vcap6
  uid: eceb1d9a-3d9a-4741-9f40-0f4ac23df5f7
spec:
  template:
    metadata:
      annotations:
        autoscaling.knative.dev/minScale: "1"
      creationTimestamp: null
    spec:
      containerConcurrency: 0
      containers:
      - env:
        - name: ServiceBindingOperatorChangeTriggerEnvVar
          value: "2020-04-23T09:18:32Z"
        envFrom:
        - secretRef:
            name: binding-request-vcap6
        image: docker.io/qibobo/go4translator_vcap:latest
        name: user-container
        ports:
        - containerPort: 8080
        readinessProbe:
          successThreshold: 1
          tcpSocket:
            port: 0
        resources: {}
      timeoutSeconds: 300
  traffic:
  - latestRevision: true
    percent: 100
status:
  address:
    url: http://mytranslator-web-knative-vcap6.e837ea57-f035.svc.cluster.local
  conditions:
  - lastTransitionTime: "2020-04-23T09:18:41Z"
    status: "True"
    type: ConfigurationsReady
  - lastTransitionTime: "2020-04-23T09:18:42Z"
    message: Ingress has not yet been reconciled.
    reason: IngressNotConfigured
    status: Unknown
    type: Ready
  - lastTransitionTime: "2020-04-23T09:18:42Z"
    message: Ingress has not yet been reconciled.
    reason: IngressNotConfigured
    status: Unknown
    type: RoutesReady
  latestCreatedRevisionName: mytranslator-web-knative-vcap6-xvlqh
  latestReadyRevisionName: mytranslator-web-knative-vcap6-xvlqh
  observedGeneration: 2
  traffic:
  - latestRevision: true
    percent: 100
    revisionName: mytranslator-web-knative-vcap6-xvlqh
  url: http://mytranslator-web-knative-vcap6.e837ea57-f035.dev-serving.knative.dev.appdomain.cloud
```

**3 Then I updated the kn service by kubectl apply:**

```
apiVersion: serving.knative.dev/v1
kind: Service 
metadata:
  name: mytranslator-web-knative-vcap6
spec:
  template:
    metadata:
      annotations:
        autoscaling.knative.dev/minScale: "1"
    spec:
      containers:
        - image: docker.io/qibobo/go4translator_vcap:v2
          ports:
          - containerPort: 8080
```

And then the secret is not there anymore:

```
kubectl get ksvc mytranslator-web-knative-vcap6 -o yamlapiVersion: serving.knative.dev/v1
kind: Service
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"serving.knative.dev/v1","kind":"Service","metadata":{"annotations":{},"name":"mytranslator-web-knative-vcap6","namespace":"e837ea57-f035"},"spec":{"template":{"metadata":{"annotations":{"autoscaling.knative.dev/minScale":"1"}},"spec":{"containers":[{"image":"docker.io/qibobo/go4translator_vcap:v2","ports":[{"containerPort":8080}]}]}}}}
    serving.knative.dev/creator: system:serviceaccount:e837ea57-f035:e837ea57-f035-writer
    serving.knative.dev/lastModifier: system:serviceaccount:e837ea57-f035:e837ea57-f035-writer
  creationTimestamp: "2020-04-23T09:15:28Z"
  generation: 3
  name: mytranslator-web-knative-vcap6
  namespace: e837ea57-f035
  resourceVersion: "48639848"
  selfLink: /apis/serving.knative.dev/v1/namespaces/e837ea57-f035/services/mytranslator-web-knative-vcap6
  uid: eceb1d9a-3d9a-4741-9f40-0f4ac23df5f7
spec:
  template:
    metadata:
      annotations:
        autoscaling.knative.dev/minScale: "1"
      creationTimestamp: null
    spec:
      containerConcurrency: 0
      containers:
      - image: docker.io/qibobo/go4translator_vcap:v2
        name: user-container
        ports:
        - containerPort: 8080
        readinessProbe:
          successThreshold: 1
          tcpSocket:
            port: 0
        resources: {}
      timeoutSeconds: 300
  traffic:
  - latestRevision: true
    percent: 100
status:
  address:
    url: http://mytranslator-web-knative-vcap6.e837ea57-f035.svc.cluster.local
  conditions:
  - lastTransitionTime: "2020-04-23T09:28:49Z"
    message: "Revision \"mytranslator-web-knative-vcap6-gsk6z\" failed with message:
      Container failed with: 2020/04/23 09:28:47 go4translator started.\n2020/04/23
      09:28:47 ----vcap_services string: \n2020/04/23 09:28:47 ----serviceClassName:
      language-translator\n2020/04/23 09:28:47 The credential is <nil>\npanic: runtime
      error: invalid memory address or nil pointer dereference\n[signal SIGSEGV: segmentation
      violation code=0x1 addr=0x0 pc=0x7e5fd5]\n\ngoroutine 1 [running]:\nmain.main()\n\t/app/vcap/main.go:54
      +0x115\n."
    reason: RevisionFailed
    status: "False"
    type: ConfigurationsReady
  - lastTransitionTime: "2020-04-23T09:28:49Z"
    message: "Revision \"mytranslator-web-knative-vcap6-gsk6z\" failed with message:
      Container failed with: 2020/04/23 09:28:47 go4translator started.\n2020/04/23
      09:28:47 ----vcap_services string: \n2020/04/23 09:28:47 ----serviceClassName:
      language-translator\n2020/04/23 09:28:47 The credential is <nil>\npanic: runtime
      error: invalid memory address or nil pointer dereference\n[signal SIGSEGV: segmentation
      violation code=0x1 addr=0x0 pc=0x7e5fd5]\n\ngoroutine 1 [running]:\nmain.main()\n\t/app/vcap/main.go:54
      +0x115\n."
    reason: RevisionFailed
    status: "False"
    type: Ready
  - lastTransitionTime: "2020-04-23T09:24:10Z"
    status: "True"
    type: RoutesReady
  latestCreatedRevisionName: mytranslator-web-knative-vcap6-gsk6z
  latestReadyRevisionName: mytranslator-web-knative-vcap6-xvlqh
  observedGeneration: 3
  traffic:
  - latestRevision: true
    percent: 100
    revisionName: mytranslator-web-knative-vcap6-xvlqh
  url: http://mytranslator-web-knative-vcap6.e837ea57-f035.dev-serving.knative.dev.appdomain.cloud  
```

**Comment by otaviof** [comment]
Community meeting notes:
- It's a road blocker, we need to reproduce the scenario and troubleshoot it again;

**Comment by Avni-Sharma** [comment]
an extension of the issue is also like the following scenario:
- Create an imported app
- Have a backend CR defined
- Create a SBR
edit the deployment to remove the `envFrom: secret ...` 
The operator won't reconcile to inject the secret again. [The secret exists]

**Comment by isutton** [comment]
> I think we have two options to address the reconcile issue.
> 
> 1. Add informer to watch whatever CRD defined in the applicationSelector/backingServiceSelector(s). This might cause overhead as some resources like knative service are designed to update frequently and adding too many resource kinds to watch would cause potential performance issues. It would be even worse if the service binding operator is running at cluster level.

Can you elaborate? 

I believe that between watching all resource GVKs and a subscription based watch list, the former is more helpful to the user on the long term even if the latter is simpler to implement.

**Comment by sbose78** [comment]
@cdlliuy I updated the title a bit to make it more accurate. I think I've finally caught up with this thread. The behaviour you've specified is as per design. 

1. If the backing service and it's associated secrets/configmaps change, controller should reconcile.
2. If the application workload changes, the operator doesn't reconcile because it was design decision we had made where we wanted to the workload owner to be the king. We didn't want the SBR controller to manage it at all. Imagine being an admin who purposely removes the secret finds it showing up again and again!

Having said that, it should be fairly easy to change the behaviour:
- Have the controller start watching the typical workload types ( Deployments, DeploymentConfigs, Knative ..)
- Have the reconciler annotate application workloads the way we annotate everything else - here's what had disabled it https://github.com/redhat-developer/service-binding-operator/pull/179

With a smart use of predicates & mapper we should be able to avoid the noise as well.

**Comment by DhritiShikhar** [comment]
JIRA issue -> https://issues.redhat.com/browse/APPSVC-557

**Comment by cdlliuy** [comment]
@sbose78 ,  @DhritiShikhar ,   got it.  So you will track this down and submit related changes to make this happen, right?

**Comment by int8dev** [comment]
I think the comment from @sbose78 makes total sense. I could raise PR for this if currently nobody is occupied.

**Comment by DhritiShikhar** [comment]
@int8dev Sure. Please go ahead and raise a PR for the same. I will review it.

**Comment by DhritiShikhar** [comment]
@lanceliuu 

Your PR (https://github.com/redhat-developer/service-binding-operator/pull/442) solves this scenario (https://github.com/redhat-developer/service-binding-operator/issues/422#issuecomment-618484986)

but the original issue (https://github.com/redhat-developer/service-binding-operator/issues/422#issue-599321019) is still not addressed -> https://github.com/redhat-developer/service-binding-operator/pull/442#issuecomment-621119274

Are you planning to look into the original issue as well?

**Comment by lanceliuu** [comment]
@DhritiShikhar The original issue should also be addressed if we fixed the issue I mentioned in https://github.com/redhat-developer/service-binding-operator/pull/442#issuecomment-621108140. If the sbr could enter into a fail and requeue state after deleting the workload, then it will be successfully binding to the newly created application eventually after some back-off time.

**Comment by DhritiShikhar** [comment]
@lanceliuu Yes. I can see that the original issue is solved as well. Thanks!

**Comment by sbose78** [comment]
@cdlliuy @navidsh @DhritiShikhar Could we track the issue of watching backing services in a separate issue, since we have already handled application workloads ?

**Comment by sbose78** [comment]
https://github.com/redhat-developer/service-binding-operator/issues/508
